### PR TITLE
Default implementation of image segmentation using SAM

### PR DESCRIPTION
**Pull request by v-imark**

### CODE REVIEW

**Comment by v-imark** [comment]
Default implementation of the segmentation part of the project implemented. Saves the masks in a folder and stores metadata from the output in a csv file.

File: main.py [hunk 1]
```diff
@@ -1,13 +1,19 @@
-import os
-from pathlib import Path
+import subprocess
+import sys
 
-import cv2
+command = [
```
Comment (ResearchDaniel) [inline main.py]: Use argparse
https://docs.python.org/3/library/argparse.html

File: main.py [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
 
-# load_input_data(dataset="cifar10", size=5)
-
-sample_image = cv2.imread("./outputs/sample_images/cifar10/test/7.png")
-
-image_dir = Path("outputs", "annotated_images", "test")
-os.makedirs(image_dir, exist_ok=True)
-os.chdir(image_dir)
-os.makedirs(Path("masks"), exist_ok=True)
+subprocess.run(command)
```
Comment (ResearchDaniel) [inline main.py]: Use `__main__`
https://docs.python.org/3/library/__main__.html#module-__main__

File: main.py
Comment (ResearchDaniel) [inline main.py:1]: Is the purpose of this file to supply default arguments?
That can be done via argparse

File: segment_images.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import os
+from pathlib import Path
+
+import cv2  # type: ignore
+import numpy as np
+import torch
+from segment_anything import SamAutomaticMaskGenerator, sam_model_registry
+
+parser = argparse.ArgumentParser(description="A script that segments multiple images")
+parser.add_argument("--sam_model_type", help="First argument")
```
Comment (ResearchDaniel) [inline segment_images.py]: Add supported options and default arguments to parser args

File: segment_images.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+parser.add_argument("--checkpoint", help="Second argument")
```
Comment (ResearchDaniel) [inline segment_images.py]: Adjust help text

File: segment_images.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    DEVICE = torch.device(device if torch.cuda.is_available() else "cpu")
+    sam = sam_model_registry[model_type](checkpoint=checkpoint_path)
+    sam.to(device=DEVICE)
+    mask_generator = SamAutomaticMaskGenerator(sam, points_per_side=points_per_side)
+    return mask_generator, sam, DEVICE
+
+
+def compute_masks(image, generator: SamAutomaticMaskGenerator):
+    """Segments an image into masks using SAM-model"""
+    # start_time = time.time()
```
Comment (ResearchDaniel) [inline segment_images.py]: Remove commented code

File: segment_images.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    print("Segmenting image...")
```
Comment (ResearchDaniel) [inline segment_images.py]: More informative with name of image

File: segment_images.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+    masks = generator.generate(image_rgb)
+    # end_time = time.time()
+    # elapsed_time = end_time - start_time
+    # print("Image segmentation complete.", f"Elapsed time: {elapsed_time} seconds.")
+
+    # Todo: exclude small masks
```
Comment (ResearchDaniel) [inline segment_images.py]: Not part of this function (Single responsibility principle)

File: segment_images.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    return masks
+
+
+def save_masks(masks, image, img_name, output_path):
+    masks_path = Path(output_path, img_name, "masks")
+    os.makedirs(masks_path, exist_ok=True)
+
+    count = 0
+    csv_data = []
+    for mask in masks:
```
Comment (ResearchDaniel) [inline segment_images.py]:
```suggestion
    for count, mask in enumerate(masks):
```

Comment (ResearchDaniel) [reply segment_images.py]: Is count necessary to include?

File: segment_images.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    generator, _, _ = register_sam(
+        model_type=args.sam_model_type,
+        checkpoint_path=args.checkpoint,
+        points_per_side=args.points_per_side,
+    )
+
+    img_folder = Path(args.data_path)
+
+    # Will change these later
+    output_name = f"cifar10_{args.size}"
```
Comment (ResearchDaniel) [inline segment_images.py]: Add as arguments

File: segment_images.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    output_path = Path("outputs", output_name)
+    os.makedirs(output_path, exist_ok=True)
+
+    img_folder_files = list(img_folder.iterdir())
+
+    # Loop through the image files and read them
+    for i in range(int(args.size)):
+        if str(img_folder_files[i]).lower().endswith((".png", ".jpg", ".jpeg")):
```
Comment (ResearchDaniel) [inline segment_images.py]: Use Path.suffix:
https://docs.python.org/3/library/pathlib.html
```suggestion
        if img_folder_files[i].suffix in [".png", ".jpg", ".jpeg"]:
```

File: segmentation.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+from pathlib import Path
+
+import cv2
+import torch
+from segment_anything import SamAutomaticMaskGenerator, sam_model_registry
+
+sample_image = cv2.imread("./outputs/sample_images/cifar10/test/7.png")
+
+
+def register_sam(
```
Comment (ResearchDaniel) [inline segmentation.py]: Duplicate code

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]
I also recommend adding linter and github actions to aid in code formating
https://github.com/pre-commit/pre-commit

File: .idea/inspectionProfiles/Project_Default.xml
Comment (ResearchDaniel) [inline .idea/inspectionProfiles/Project_Default.xml:1]: Remove .ideas folder and add to git ignore file

File: segment_images.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    DEVICE = torch.device(device if torch.cuda.is_available() else "cpu")
+    sam = sam_model_registry[model_type](checkpoint=checkpoint_path)
+    sam.to(device=DEVICE)
+    mask_generator = SamAutomaticMaskGenerator(sam, points_per_side=points_per_side)
+    return mask_generator, sam, DEVICE
+
+
+def compute_masks(image, name, generator: SamAutomaticMaskGenerator):
+    """Segments an image into masks using SAM-model"""
+    print(f"Segmenting image {name}...")
```
Comment (ResearchDaniel) [inline segment_images.py]: Would move this to the calling side. Unnecessary to specify name just for printing

**Review by ResearchDaniel — APPROVED** [review]
